File: src/main/java/com/zepben/vertxutils/routing/handlers/params/QueryParamRule.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing.handlers.params;

import com.zepben.annotations.EverythingIsNonnullByDefault;

import javax.annotation.Nullable;

@EverythingIsNonnullByDefault
public class QueryParamRule<T> extends ParamRule<String, T> {

    @Nullable private final T defaultValue;
    private final boolean isRequired;

    public static <T> QueryParamRule<T> of(String name, RequestValueConverter<String, T> converter) {
        return new QueryParamRule<>(name, converter, null, false);
    }

    public static <T> QueryParamRule<T> of(String name, RequestValueConverter<String, T> converter, T defaultValue) {
        return new QueryParamRule<>(name, converter, defaultValue, false);
    }

    public static <T> QueryParamRule<T> ofRequired(String name, RequestValueConverter<String, T> converter) {
        return new QueryParamRule<>(name, converter, null, true);
    }

    private QueryParamRule(String name, RequestValueConverter<String, T> converter, @Nullable T defaultValue, boolean isRequired) {
        super(name, converter);
        this.defaultValue = defaultValue;
        this.isRequired = isRequired;
    }

    @Nullable
    public T defaultValue() {
        return defaultValue;
    }

    public boolean isRequired() {
        return isRequired;
    }
}

File: src/main/java/com/zepben/vertxutils/routing/handlers/params/QueryParamRule.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing.handlers.params

class QueryParamRule<T> private constructor(
    name: String,
    converter: RequestValueConverter<String, T>,
    val defaultValue: T?,
    val isRequired: Boolean,
) : ParamRule<String, T>(name, converter) {

    companion object {

        fun <T> of(name: String, converter: RequestValueConverter<String, T>): QueryParamRule<T> =
            QueryParamRule(name, converter, null, false)

        fun <T> of(name: String, converter: RequestValueConverter<String, T>, defaultValue: T): QueryParamRule<T> =
            QueryParamRule(name, converter, defaultValue, false)

        fun <T> ofRequired(name: String, converter: RequestValueConverter<String, T>): QueryParamRule<T> =
            QueryParamRule(name, converter, null, true)
    }

}
